07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
+			auto imageAllocation = m_device->allocate(images[imageIdx]->getMemoryReqs(), images[imageIdx].get(), IDeviceMemoryAllocation::EMAF_NONE);
+
+			if (!imageAllocation.isValid())
+				logFailAndTerminate("Failed to allocate Device Memory compatible with our GPU Buffer!\n");
+
+			
+			IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imageLayoutTransitionBarrier0; // TODO: better names maybe?
+			IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imageLayoutTransitionBarrier1;
+			{
+				IImage::SSubresourceRange imgSubresourceRange{};
+				imgSubresourceRange.aspectMask = IImage::E_ASPECT_FLAGS::EAF_COLOR_BIT;
+				imgSubresourceRange.baseMipLevel = 0u;
+				imgSubresourceRange.baseArrayLayer = 0u;
+				imgSubresourceRange.levelCount = 1;
+				imgSubresourceRange.layerCount = 1u;
+
+				imageLayoutTransitionBarrier0.barrier.dep.srcAccessMask = ACCESS_FLAGS::NONE;
+				imageLayoutTransitionBarrier0.barrier.dep.dstAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS;
+				imageLayoutTransitionBarrier0.barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::NONE;
+				imageLayoutTransitionBarrier0.barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS; //TODO: COPY_BIT?
+				imageLayoutTransitionBarrier0.oldLayout = asset::IImage::LAYOUT::UNDEFINED;
+				imageLayoutTransitionBarrier0.newLayout = asset::IImage::LAYOUT::TRANSFER_DST_OPTIMAL; // TODO: use more suitable layout
+				imageLayoutTransitionBarrier0.image = images[imageIdx].get();
+				imageLayoutTransitionBarrier0.subresourceRange = imgSubresourceRange;
+
+				imageLayoutTransitionBarrier1.barrier.dep.srcAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS; 
+				imageLayoutTransitionBarrier1.barrier.dep.dstAccessMask = ACCESS_FLAGS::NONE;
+				imageLayoutTransitionBarrier1.barrier.dep.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS; //TODO: COPY_BIT?
+				imageLayoutTransitionBarrier1.barrier.dep.dstStageMask = PIPELINE_STAGE_FLAGS::NONE;
+				imageLayoutTransitionBarrier1.oldLayout = asset::IImage::LAYOUT::TRANSFER_DST_OPTIMAL;
+				imageLayoutTransitionBarrier1.newLayout = asset::IImage::LAYOUT::GENERAL; // TODO: use more suitable layout
+				imageLayoutTransitionBarrier1.image = images[imageIdx].get();
+				imageLayoutTransitionBarrier1.subresourceRange = imgSubresourceRange;
+			}
+
+			core::smart_refctd_ptr<ISemaphore> imgFillSemaphore = m_device->createSemaphore(0); // TODO: don't create semaphore every iteration
+			IQueue::SSubmitInfo::SCommandBufferInfo imgFillCmdBuffInfo = { cmdBuff.get() };
+
+			IQueue::SSubmitInfo::SSemaphoreInfo imgFillSemaphoreWaitInfo = {
+				.semaphore = imgFillSemaphore.get(),
+				.value = 1,
+				.stageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS
+			};
+
+			IQueue::SSubmitInfo imgFillSubmitInfo = {
+				.waitSemaphores = {&imgFillSemaphoreWaitInfo, 1},
+				.commandBuffers = {&imgFillCmdBuffInfo, 1}
+			};
+
+			transferUpQueue->submit({ &imgFillSubmitInfo, 1 });
```

Comment:
why are you doing a submit of an empty (freshly reset commandbuffer), just to signal a temporary semaphore to 1 !?

Reply:
especially since that your `intendedSubmit` uses a `value` of 0 for its scratch semaphore